core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java

```diff
@@ -101,24 +107,70 @@ public static BlockchainSourceImpl getInstance() {
 	}
 
 	@Override
-	public void createAccount() throws BlockchainException {
-		createKinAccountIfNeeded();
+	public void loadAccount(String kinUserId) throws BlockchainException {
+		migrateToMultipleUsers(kinUserId);
+		createOrLoadAccount(kinUserId);
 		initBalance();
 	}
 
-	private void createKinAccountIfNeeded() throws BlockchainException {
-		if (kinClient.hasAccount()) {
+	/**
+	 * Support backward compatibility,
+	 * load the old account and migrate to current multiple users implementation.
+	 */
+	private void migrateToMultipleUsers(String kinUserId) {
+		final boolean isMigrated = local.getIsMigrated();
+		if (!isMigrated && kinClient.hasAccount()) {
+			final int accountIndex = local.getAccountIndex();
+			KinAccount account;
+			if(accountIndex == NOT_EXIST) {
+				account = kinClient.getAccount(0);
+			} else {
+				Logger.log(new Log().withTag(TAG).put("migrateToMultipleUsers user:", kinUserId));
+				account = kinClient.getAccount(accountIndex);
+				local.deleteAccountIndexKey();
+			}
+			local.setDidMigrate();
+			local.setActiveUserWallet(kinUserId, account.getPublicAddress());
+		}
+	}
+
+	private void createOrLoadAccount(String kinUserId) throws BlockchainException {
+		final String lastWalletAddress = local.getLastWalletAddress(kinUserId);
+		if (kinClient.hasAccount() && !StringUtil.isEmpty(lastWalletAddress)) {
+			Logger.log(new Log().withTag(TAG).text("createOrLoadAccount").put("currentUserId", currentUserId)
+				.put("kinUserId", kinUserId));
+
+			// Match between last wallet address to wallets on device.
+			if (!StringUtil.isEmpty(lastWalletAddress)) {
+				for (int i = 0; i < kinClient.getAccountCount(); i++) {
+					KinAccount account = kinClient.getAccount(i);
+					if (lastWalletAddress.equals(account.getPublicAddress())) {
+						this.account = account;
+						break;
+					}
+				}
+			}
+
+			// No matching found
 			if (account == null) {
-				account = kinClient.getAccount(activeAccountIndex);
+				account = createAccount();
```

Comment:
this should never happens if we have lastWalletAddress we must have a matching wallet

Reply:
i would keep it there for safty

Reply:
better know we have a problem then overriding  users wallet